### PR TITLE
Support `spill` mechanism for v6d server

### LINKED ISSUE #740: Spilling: dump cold blobs when memory is full in Allocation.
Describe your problem
---------------------

Support `spill` and `reload` functionality for blobs. Intuitively we can store a `BlobFault` flag for the cold blobs in our memory storage, and trigger the reload callback when we touch them.

### PR DESCRIPTION
**Pull request by ZjuYTW**
Signed-off-by: ZjuYTW <wangyitao999@outlook.com>

What do these changes do?
-------------------------

Based on the last PR which implemented a simple LRU, I will support `Spill` in this one.

Design Breif
------------
To achive the spill functionality, we should
1. Detect OOM
2. Sieve out  the cold objects, which supported by LRU.
3. Spill the objects onto disks, to make memory avaliable.
4. Design a mechanism to fetch back the data

Related new data structures:
* `spilling_threshold` memory usage normal threshold
* `spilled_objs_`,  a hash map that records all the spilled obj
*  `spilled_path`, the target path we gonna spill data onto.

Detailed Design
----------------

### Detect OOM
When we call `BulkStore::Create`, if `AllocateMemory` returns nullptr, means we are currently running out of memory.

### Sieve out cold objects
Add a new method inside of `LRU`, like `SpillColdObj(size_t sz)` to spill the least used objects till their size sums up to `sz`.

### Dump onto disk
A currently plan is to make one object one spilled file, and we could just trace them using the format `tmp_spilled_<object_id>`. 
As for the content, we should dump 1. `object_id` 2. `checksum` (optional) 3. `data` . All of them should follow the format of `|Fixed 64 bits length| + |Length's content|`

### Object fault(callback)
Maybe we could implement this inside of `Payload`, by a flag `is_spilled`.  Whenever the payload is touched by its methods, it will check out the flag and if its `data` is needed, do the fetch operation.

Here is one thing that in my understanding, whenever the data is touched, the object should be kicked out of our lru. Thus maybe we should also associcate `BulkStore::RemoveFromColdList` with the callback?

### Miscellaneous
1. The Spill method should be done only once, although the concurrent executions of `CreateBuffer`. Because in my current design, we should specify the `sz` to `SpillColdObj` and we don't want to spill multiply `sz` of memory.

So maybe a locked is needed here, or some kind of synchronizing singal here.

2. Regarding to the spilled file, can we use json format or the format I mentioned above?

3. In the `sepc_` we config the memory limitation and do we need another config to specify the `memory_spilling_threshold`, or just some portion of memory like `1/3 * memory_limit`

Related issue number
--------------------

Fixes #740 

cc @septicmk

### CODE REVIEW

**Comment by ZjuYTW** [comment]
Seems something wrong with `apt-get`, anyone could help me with this? T_T, not so familiar with CI workflow

File: src/server/memory/usage.h [hunk 1]
```diff
@@ -219,26 +227,68 @@ class ColdObjectTracker
...
       std::unique_lock<decltype(mu_)> locked;
       auto it = map_.find(id);
       if (it == map_.end()) {
+        auto it = spilled_obj_.find(id);
+        if (it == spilled_obj_.end()) {
+          return Status::OK();
+        }
+        if (!fast_delete) {
+          RETURN_ON_ERROR(store_ptr->ReloadPayload_(id, it->second));
+        }
```
Comment (septicmk) [inline src/server/memory/usage.h:250]: What will happen, if fast_delete is true?

File: src/server/util/spill_file.cc [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+Status SpillWriteFile::Sync() {
+  RETURN_ON_ERROR(io_adaptor_->Flush());
+  io_adaptor_ = nullptr;
+  return Status::OK();
+}
+
+void SpillReadFile::Init(uint64_t object_id) {
+  if (io_adaptor_) {
+    LOG(WARNING) << "Warning: may not flushed before io_adaptor_ is overwriten";
+  }
```
Comment (septicmk) [inline src/server/util/spill_file.cc:76]: ditto

File: src/server/memory/usage.h [hunk 3]
```diff
@@ -293,6 +351,8 @@ class ColdObjectTracker
   Status MarkAsCold(ID const& id, std::shared_ptr<P> payload) {
     if (payload->IsSealed()) {
       cold_obj_lru_.Ref(id, payload);
+    } else {
+      LOG(WARNING) << "Not Sealed, Object ID: " << id;
```
Comment (septicmk) [inline src/server/memory/usage.h:355]: Return a Status and handle it outside gracefully.

File: src/server/util/spill_file.cc [hunk 4]
```diff
@@ -0,0 +1,117 @@
...
+#include "common/util/uuid.h"
+#include "io/io/local_io_adaptor.h"
+#include "server/memory/memory.h"
+#include "server/util/file_io_adaptor.h"
+
+namespace util {
+using vineyard::BulkStore;
+using vineyard::Payload;
+using vineyard::Status;
+void PutFixed32(std::string* dst, uint32_t value) {
```
Comment (septicmk) [inline src/server/util/spill_file.cc:33]: Is this an unused function?

File: src/server/server/vineyard_server.cc [hunk 5]
```diff
@@ -107,9 +107,13 @@ Status VineyardServer::Serve(StoreType const& bulk_store_type) {
...
   } else if (bulk_store_type_ == StoreType::kDefault) {
     bulk_store_ = std::make_shared<BulkStore>();
-    RETURN_ON_ERROR(bulk_store_->PreAllocate(
-        spec_["bulkstore_spec"]["memory_size"].get<size_t>()));
-
+    auto mem_limit = spec_["bulkstore_spec"]["memory_size"].get<size_t>();
+    RETURN_ON_ERROR(bulk_store_->PreAllocate(mem_limit));
+    // we should set the threshold as a portion of mem limit
+    // TODO: Maybe someday we should make SpillSize a configurable option
+    bulk_store_->SetMemSpillSize(mem_limit / 3);
```
Comment (septicmk) [inline src/server/server/vineyard_server.cc:114]: Make the spilling threshold a configurable value. For example, a spilling rate r (0<=r<1), and trigger spilling when used_memory < (mem_limit * r)

File: CMakeLists.txt [hunk 6]
```diff
@@ -558,7 +559,9 @@ if(BUILD_VINEYARD_SERVER)
                                            ${OPENSSL_LIBRARIES}
                                            nlohmann_json::nlohmann_json
                                            TBB::tbb
+                                           vineyard_io
```
Comment (septicmk) [inline CMakeLists.txt:562]: do we really need vineyard_io here? could we use arrow to replace this?

File: src/server/memory/usage.h (hunk 1 above)
Comment (ZjuYTW) [reply src/server/memory/usage.h:250]: `fast_delete` is used to denote the situation that, if the user called `delete`-liked rpc/ipc to the server, then we shouldn't reload the file from disk, instead we should directly remove the payload.

File: src/server/util/spill_file.cc (hunk 4 above)
Comment (ZjuYTW) [reply src/server/util/spill_file.cc:33]: yes, maybe we could keep it for later use...?

File: src/server/util/file_io_adaptor.cc [hunk 7]
```diff
@@ -0,0 +1,125 @@
...
+Status FileIOAdaptor::DeleteDir() {
+  return Status::ArrowError(fs_->DeleteDir(location_));
+}
+
+Status FileIOAdaptor::RemoveFiles(const std::vector<std::string>& paths) {
+  return Status::ArrowError(fs_->DeleteFiles(paths));
+}
+
+Status FileIOAdaptor::RemoveFile(const std::string& path) {
+  // FIXME: here will encounter a double-free bug
```
Comment (ZjuYTW) [inline src/server/util/file_io_adaptor.cc:121]: @septicmk  please have a look, can't Remove File in `SpillReadFile::Read()`

File: src/server/memory/usage.h (hunk 1 above)
Comment (ZjuYTW) [reply src/server/memory/usage.h:253]: When I want to add the delete file logic in [here](https://github.com/v6d-io/v6d/pull/815/commits/3439b400171306cc7de3cab0a8c52ecfa1dc2632#diff-d7ef7d2035b2b3d576bc1a4f04bf065eeab33c3f8baf4c9fcb3e0f55c6d1eec6R114), the program complains for double free

File: src/server/util/spec_resolvers.cc [hunk 8]
```diff
@@ -95,6 +99,11 @@ json BulkstoreSpecResolver::resolve() const {
   size_t bulkstore_limit = parseMemoryLimit(FLAGS_size);
   spec["memory_size"] = bulkstore_limit;
   spec["stream_threshold"] = FLAGS_stream_threshold;
+  spec["spill_path"] = FLAGS_spill_path;
+  spec["spill_low_bound_rate"] = FLAGS_spill_low_rate;
+  spec["spill_up_bound_rate"] = FLAGS_spill_up_rate;
```
Comment (septicmk) [inline src/server/util/spec_resolvers.cc:104]: `spill_lower_bound_rate` and `spill_upper_bound_rate`

File: src/server/memory/usage.h [hunk 9]
```diff
@@ -221,26 +228,70 @@ class ColdObjectTracker
...
+    Status Spill(size_t sz, std::shared_ptr<Der> bulk_store_ptr) {
       std::unique_lock<decltype(mu_)> locked;
-      if (list_.empty()) {
-        return {Status::Invalid(), {-1, nullptr}};
+      size_t spilled_sz = 0;
+      auto st = Status::OK();
+      auto it = list_.rbegin();
+      while (it != list_.rend()) {
+        st = bulk_store_ptr->SpillPayload_(it->second);
+        // st = it->second->Spill();
```
Comment (septicmk) [inline src/server/memory/usage.h:269]: Clean up unnecessary comments.

File: src/server/util/file_io_adaptor.cc [hunk 10]
```diff
@@ -0,0 +1,124 @@
...
+
+#include "arrow/filesystem/filesystem.h"
+#include "basic/ds/arrow_utils.h"
+
+namespace util {
+using vineyard::Status;
+
+FileIOAdaptor::FileIOAdaptor(const std::string& location) {
+  // TODO(ZjuYTW): Maybe we should check the validation of dir_path
+  location_ = location;
```
Comment (septicmk) [inline src/server/util/file_io_adaptor.cc:27]: Yes, we should check the location, and handle the error properly (show warning and disable spilling).

File: src/server/util/spec_resolvers.cc [hunk 11]
```diff
@@ -95,6 +99,11 @@ json BulkstoreSpecResolver::resolve() const {
   size_t bulkstore_limit = parseMemoryLimit(FLAGS_size);
   spec["memory_size"] = bulkstore_limit;
   spec["stream_threshold"] = FLAGS_stream_threshold;
+  spec["spill_path"] = FLAGS_spill_path;
+  spec["spill_lower_bound_rate"] = FLAGS_spill_lower_rate;
+  spec["spill_upper_bound_rate"] = FLAGS_spill_upper_rate;
+  // if need to config the spill threshold, add json option here
+  // spec["memory_spill_threshold"]
```
Comment (septicmk) [inline src/server/util/spec_resolvers.cc:106]: Clean up unnecessary comments.

**Comment by septicmk** [comment]
LGTM.

**Review by septicmk — APPROVED** [review]

File: CMakeLists.txt [hunk 12]
```diff
@@ -541,6 +541,7 @@ include("cmake/GenerateVineyardJava.cmake")
 
 # build vineyardd
 if(BUILD_VINEYARD_SERVER)
+    set(BUILD_VINEYARD_IO)
```
Comment (sighingnow) [inline CMakeLists.txt:544]: Not used, I think.

File: CMakeLists.txt [hunk 13]
```diff
@@ -559,6 +560,13 @@ if(BUILD_VINEYARD_SERVER)
                                            nlohmann_json::nlohmann_json
                                            TBB::tbb
     )
+    if(ARROW_SHARED_LIB)
+        target_link_libraries(vineyardd PUBLIC ${ARROW_SHARED_LIB})
+    else()
+        target_link_libraries(vineyardd PUBLIC ${ARROW_STATIC_LIB})
+    endif()
+
+    target_include_directories(vineyardd PRIVATE $<BUILD_INTERFACE:${CMAKE_CURRENT_SOURCE_DIR}/thirdparty/ctti/include>)
```
Comment (sighingnow) [inline CMakeLists.txt:569]: Why requires ctti in the vineyardd?

File: src/common/memory/payload.cc [hunk 14]
```diff
@@ -14,8 +14,9 @@ limitations under the License.
 */
 
 #include "common/memory/payload.h"
-
 #include <cstdint>
+#include <string>
+#include "common/util/logging.h"
```
Comment (sighingnow) [inline src/common/memory/payload.cc:19]: Group the include lines, put `#include "common/util/logging.h"` to the group of "common/..." includes above.

File: src/server/async/socket_server.cc [hunk 15]
```diff
@@ -1396,6 +1399,19 @@ bool SocketConnection::doIsInUse(json const& root) {
   return false;
 }
 
+bool SocketConnection::doIsSpilled(json const& root) {
+  auto self(shared_from_this());
+  ObjectID id;  // Must be a blob id.
+  TRY_READ_REQUEST(ReadIsSpilledRequest, root, id);
+  bool is_spilled = false;
+  RESPONSE_ON_ERROR(
+      server_ptr_->GetBulkStore<ObjectID>()->IsSpilled(id, is_spilled));
```
Comment (sighingnow) [inline src/server/async/socket_server.cc:1408]: Use `bulk_store_` instead.

File: src/server/util/spec_resolvers.cc [hunk 16]
```diff
@@ -32,6 +32,10 @@ DEFINE_string(meta, "etcd", "Metadata storage, can be one of: etcd, local");
 DEFINE_string(etcd_endpoint, "http://127.0.0.1:2379", "endpoint of etcd");
 DEFINE_string(etcd_prefix, "vineyard", "path prefix in etcd");
 DEFINE_string(etcd_cmd, "", "path of etcd executable");
+DEFINE_string(spill_path, "/tmp/spill_path/",
+              "path of spilling temporary files");
```
Comment (sighingnow) [inline src/server/util/spec_resolvers.cc:36]: By default the spill shouldn't be enabled so please change the default value to `""`.

File: src/server/memory/usage.h [hunk 17]
```diff
@@ -320,9 +438,44 @@ class ColdObjectTracker
...
+    util::SpillReadFile read_file(spill_path_);
+    RETURN_ON_ERROR(read_file.Read(payload, Self().shared_from_this()));
+    return Status::OK();
+  }
+
+  Status DeletePayloadFile_(const ID& id) {
+    util::FileIOAdaptor io_adaptor(spill_path_);
+    RETURN_ON_ERROR(io_adaptor.RemoveFile(spill_path_ + std::to_string(id)));
+    return Status::OK();
+  }
```
Comment (sighingnow) [inline src/server/memory/usage.h:465]: No need for `_` suffix for protected member functions.

File: src/server/util/file_io_adaptor.cc [hunk 18]
```diff
@@ -0,0 +1,124 @@
...
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+#include "server/util/file_io_adaptor.h"
+#include <vector>
+
+#include "arrow/filesystem/filesystem.h"
+#include "basic/ds/arrow_utils.h"
```
Comment (sighingnow) [inline src/server/util/file_io_adaptor.cc:20]: Please move the arrow related macros to a new file `common/util/arrow.h` then we could avoid include `basic/ds/arrow_utils.h` in vineyardd.

File: src/server/util/spill_file.cc [hunk 19]
```diff
@@ -0,0 +1,126 @@
+/** Copyright 2020-2021 Alibaba Group Holding Limited.
```
Comment (sighingnow) [inline src/server/util/spill_file.cc:1]: `2020-2022` for new files.

File: src/server/util/spill_file.h [hunk 20]
```diff
@@ -0,0 +1,113 @@
...
+#include <memory>
+#include <string>
+#include "common/memory/payload.h"
+#include "common/util/status.h"
+#include "common/util/uuid.h"
+#include "io/io/i_io_adaptor.h"
+#include "io/io/io_factory.h"
+#include "io/io/local_io_adaptor.h"
+#include "server/util/file_io_adaptor.h"
+
```
Comment (sighingnow) [inline src/server/util/spill_file.h:30]: I suggest add some docs about the encoding of the spilled file, e.g., how many bytes for headers, etc. in this file.

File: test/spill_test.cc [hunk 21]
```diff
@@ -0,0 +1,173 @@
...
+#include "client/client.h"
+#include "client/ds/object_meta.h"
+#include "common/util/logging.h"
+#include "common/util/status.h"
+#include "common/util/uuid.h"
+
+using namespace vineyard;  // NOLINT
+using namespace std;       // NOLINT
+
+double delta = 1E-10;
```
Comment (sighingnow) [inline test/spill_test.cc:34]: constexpr.

**Review by sighingnow — APPROVED** [review]

File: CMakeLists.txt (hunk 13 above)
Comment (ZjuYTW) [reply CMakeLists.txt:569]:
```
[ 65%] Building CXX object CMakeFiles/vineyardd.dir/src/server/async/ipc_server.cc.o
In file included from /root/Desktop/v6d/v6d/src/client/ds/object_factory.h:27,
                 from /root/Desktop/v6d/v6d/src/client/ds/i_object.h:23,
                 from /root/Desktop/v6d/v6d/src/client/client.h:32,
                 from /root/Desktop/v6d/v6d/modules/io/io/io_factory.h:25,
                 from /root/Desktop/v6d/v6d/src/server/util/spill_file.h:27,
                 from /root/Desktop/v6d/v6d/src/server/memory/usage.h:32,
                 from /root/Desktop/v6d/v6d/src/server/memory/memory.h:46,
                 from /root/Desktop/v6d/v6d/src/server/async/ipc_server.h:29,
                 from /root/Desktop/v6d/v6d/src/server/async/ipc_server.cc:16:
/root/Desktop/v6d/v6d/src/common/util/typename.h:33:10: fatal error: ctti/detail/name_filters.hpp: No such file or directory
   33 | #include "ctti/detail/name_filters.hpp"
      |          ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```
Without ctti, the compile this complain about this

Comment (sighingnow) [reply CMakeLists.txt:569]: You don't need to include the following headers in your `spill_file.h`,

```
#include "io/io/i_io_adaptor.h"
#include "io/io/io_factory.h"
#include "io/io/local_io_adaptor.h"
```

As you already have a file io adaptor in your server and you don't use anything from the client side.

**Comment by ZjuYTW** [comment]
emmm Maybe we could skip this python analysis? @sighingnow

File: src/server/memory/usage.h [hunk 22]
```diff
@@ -16,15 +16,20 @@ limitations under the License.
 #ifndef SRC_SERVER_MEMORY_USAGE_H_
 #define SRC_SERVER_MEMORY_USAGE_H_
 
+#include <glog/logging.h>
```
Comment (sighingnow) [inline src/server/memory/usage.h:19]: Move the glog inclusion to the line under "#include "boost/....".

File: src/server/util/file_io_adaptor.h [hunk 23]
```diff
@@ -0,0 +1,70 @@
...
+limitations under the License.
+
+The implementation of `get_rss`, `get_shared_rss` and `get_peek_rss`
+are referred from https://stackoverflow.com/a/14927379/5080177, which
+has the following license header originally:
+
+  * Author:  David Robert Nadeau
+  * Site:    http://NadeauSoftware.com/
+  * License: Creative Commons Attribution 3.0 Unported License
+  *          http://creativecommons.org/licenses/by/3.0/deed.en_US
```
Comment (sighingnow) [inline src/server/util/file_io_adaptor.h:22]: There's no `get_rss`, ... in this file so please double check the copyright header if you were doing some copy-paste jobs.

File: src/server/util/spill_file.h [hunk 24]
```diff
@@ -0,0 +1,117 @@
...
+*/
+
+#ifndef SRC_SERVER_UTIL_SPILL_FILE_H_
+#define SRC_SERVER_UTIL_SPILL_FILE_H_
+
+#include <cstddef>
+#include <cstdint>
+#include <memory>
+#include <string>
+#include "common/memory/payload.h"
```
Comment (sighingnow) [inline src/server/util/spill_file.h:23]: Leave a blank line between the stdandard headers and internal headers.

**Comment by sighingnow** [comment]
> Details

I have left some comments about the code style. We can land this pull request once they got addressed.

I would take care the codacy analysis issue in another pull requests.

File: src/client/client.cc [hunk 25]
```diff
@@ -897,6 +897,23 @@ bool Client::IsInUse(ObjectID const& id) {
   return is_in_use;
 }
 
+bool Client::IsSpilled(ObjectID const& id) {
+  if (!this->connected_) {
+    VINEYARD_CHECK_OK(Status::ConnectionError("Client is not connected"));
+  }
+  std::lock_guard<std::recursive_mutex> __guard(this->client_mutex_);
```
[the remark this answers is not in the document]
Comment (ZjuYTW) [reply src/client/client.cc:904]: So I change the return type of `IsSpilled` and `IsInUse` to Status?

[the remark this answers is not in the document]
Comment (sighingnow) [reply src/client/client.cc:904]: I suggest return `Status` and use a `bool &` to return the result.

File: src/common/util/arrow.h [hunk 26]
```diff
@@ -0,0 +1,31 @@
...
+#define RETURN_ON_ARROW_ERROR_AND_ASSIGN(lhs, expr)           \
+  do {                                                        \
+    auto result = (expr);                                     \
+    if (!result.status().ok()) {                              \
+      return ::vineyard::Status::ArrowError(result.status()); \
+    }                                                         \
+    lhs = std::move(result).ValueOrDie();                     \
+  } while (0)
+}  // namespace vineyard
+
```
Comment (sighingnow) [inline src/common/util/arrow.h:30]: Could you please remove the same macro definition in `modules/basic/ds/arrow_utils.h` and include this header in that file?

The two macro looks identical and there's no reason for such a duplication.

File: src/server/memory/usage.h [hunk 27]
```diff
@@ -16,15 +16,19 @@ limitations under the License.
...
 #include <memory>
+#include <mutex>
 #include <set>
 #include <shared_mutex>
+#include <string>
 #include <unordered_map>
 #include <unordered_set>
 #include <utility>
 #include <vector>
+#include "server/util/spill_file.h"
```
Comment (sighingnow) [inline src/server/memory/usage.h:31]: Please group the includes and put this line under `#include "server/util/file_io_adaptor.h"`.

File: src/server/memory/usage.h [hunk 28]
```diff
@@ -320,9 +444,56 @@ class ColdObjectTracker
...
+    }
+    if (spill_path.back() != '/') {
+      this->spill_path.push_back('/');
+    }
+    util::FileIOAdaptor io_adaptor(this->spill_path + "test");
+    if (io_adaptor.Open("w").ok()) {
+      io_adaptor.RemoveFile(this->spill_path + "test");
+    } else {
+      LOG(ERROR) << "No such directory " << this->spill_path;
+      LOG(ERROR) << "So disable spill...";
```
Comment (sighingnow) [inline src/server/memory/usage.h:487]: Why this message is an "ERROR"?

File: src/server/memory/usage.h [hunk 29]
```diff
@@ -320,9 +444,56 @@ class ColdObjectTracker
...
+  Status DeletePayloadFile_(const ID& id) {
+    util::FileIOAdaptor io_adaptor(spill_path);
+    RETURN_ON_ERROR(io_adaptor.RemoveFile(spill_path + std::to_string(id)));
+    return Status::OK();
+  }
+
+  void SetSpillPath(const std::string& spill_path) {
+    this->spill_path = spill_path;
+    if (spill_path.empty()) {
+      LOG(ERROR) << "No spill path set, disable spill...";
```
Comment (sighingnow) [inline src/server/memory/usage.h:476]: Shouldn't be `ERROR`, I guess.

File: src/server/memory/usage.h [hunk 30]
```diff
@@ -320,9 +444,56 @@ class ColdObjectTracker
...
+  }
+
+  Status ReloadPayload_(const ID& id, std::shared_ptr<P>& payload) {
+    assert(payload->is_spilled == true);
+    util::SpillReadFile read_file(spill_path);
+    RETURN_ON_ERROR(read_file.Read(payload, Self().shared_from_this()));
+    return Status::OK();
+  }
+
+  Status DeletePayloadFile_(const ID& id) {
```
Comment (sighingnow) [inline src/server/memory/usage.h:467]: No need for the `_` suffix, seems that the previous comments hasn't been addressed.

**Review by sighingnow — APPROVED** [review]
LGTM to me. Just some minor comments.

File: src/server/memory/usage.h (hunk 28 above)
Comment (ZjuYTW) [reply src/server/memory/usage.h:487]: Cuz user enter a wrong filepath, they expect to enable spill but we couldn't do that.

Comment (ZjuYTW) [reply src/server/memory/usage.h:487]: Or maybe i could change it to WARNING

Comment (sighingnow) [reply src/server/memory/usage.h:487]: As you doesn't abort, a `WARNING` might be more accurate. I suggest

```
LOG(WARNING) << "Disabling spilling as the specified spill directory doesn't exist, or vineyardd doesn't have the permission to write it".
```

**Pull request merged**